genie-web/src/main/java/com/netflix/genie/web/services/impl/JobResolverServiceImpl.java

```diff
@@ -672,4 +724,214 @@ private int resolveJobMemory(final JobRequest jobRequest, final Command command)
+                        // the db query as a join with a subquery.
+                        mergedCriterion = this.mergeCriteria(commandClusterCriterion, jobClusterCriterion);
+                    } catch (final IllegalArgumentException e) {
+                        log.debug(
+                            "Unable to merge command cluster criterion {} and job cluster criterion {}. Skipping.",
+                            commandClusterCriterion,
+                            jobClusterCriterion,
+                            e
+                        );
+                        // Skip and move to the next combination
+                        continue;
+                    }
+
+                    queryCount++;
+                    final Set<Cluster> clusters
+                        = this.clusterPersistenceService.findClustersMatchingCriterion(mergedCriterion);
+                    if (clusters.isEmpty()) {
+                        log.debug("No clusters found for {}", mergedCriterion);
+                        this.noClusterFoundCounter.increment();
+                    } else if (clusters.size() == 1) {
+                        log.debug("Found single cluster for {}", mergedCriterion);
+                        cluster = clusters.stream().findFirst().orElse(null);
+                    } else {
+                        log.debug("Found {} clusters for {}", clusters.size(), mergedCriterion);
+                        cluster = this.selectClusterUsingClusterSelectors(
+                            Sets.newHashSet(),
+                            clusters,
+                            jobId,
+                            jobRequest
+                        );
+                    }
+                }
+
+                if (cluster != null) {
+                    break;
+                }
+            }
+
+            if (cluster == null) {
+                this.noClusterSelectedCounter.increment();
+                throw new GenieJobResolutionException("No cluster selected given criteria for job " + jobId);
+            }
+
+            log.debug("Selected cluster {} for job {}", cluster.getId(), jobId);
+
+            MetricsUtils.addSuccessTags(tags);
+            tags.add(Tag.of(MetricsConstants.TagKeys.CLUSTER_ID, cluster.getId()));
+            tags.add(Tag.of(MetricsConstants.TagKeys.CLUSTER_NAME, cluster.getMetadata().getName()));
+            return cluster;
+        } catch (final Throwable t) {
+            MetricsUtils.addFailureTagsWithException(tags, t);
+            if (t instanceof GenieJobResolutionException) {
+                throw (GenieJobResolutionException) t;
+            } else {
+                throw new GenieJobResolutionException(t);
+            }
+        } finally {
+            this.registry
+                .counter(RESOLVE_CLUSTER_CRITERIA_COMBINATION_COUNTER, tags)
+                .increment(criteriaCombinationCount);
+            this.registry
```

Comment:
The 2 counter metrics are a little weird.
The number is in the end the number of attempts made.
However they are grouped/tagged by cluster if successful.

So in the end i cannot think of a way to extract things like average/max/min number of attempts.
It's just a global counter with some tags that i cannot think of a use for.

Let's talk in person, it's easier.